### ISSUE TITLE
Trainer doesn't finish at last step.

### ISSUE THREAD

**Issue by roperi** [bghira/SimpleTuner#60]
Version (branch): data_backend

Once trainer finishes going through all the total steps, it continues with even more steps. 

```
8/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:25 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
Steps: : 624it [2:35:21, 23.78s/it, lr=1.5e-6, step_loss=0.19]  08/27/2023 13:50:28 - DEBUG - __main__ - Starting into epoch: 98
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:28 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:29 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
Steps: : 624it [2:35:25, 23.78s/it, lr=1.5e-6, step_loss=0.115]08/27/2023 13:50:32 - DEBUG - __main__ - Starting into epoch: 99
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:32 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size before rotation: (768, 768)
08/27/2023 13:50:33 - DEBUG - MultiaspectImage - Image size after rotation: (768, 768)
```

In fact, trainer continues beyond number of NUM_EPOCHS and even does validation samples every VALIDATION_STEPS. I set my  MAX_NUM_STEPS  to a incredible high number. I wonder if it's is restarting all over with MAX_NUM_STEPS.

**Comment by bghira** [comment]
the bug seemed to be that it just didn't consider that argument's value when ending training. i've never noticed that as i rely on setting the number of training steps instead.

now you can set the value for training steps to zero, and it will use epochs instead.

**Comment by bghira** [comment]
there should be more info at the start of the log on how it calculated the epochs. can you show that?

**Comment by bghira** [comment]
it does say num epochs 100. how are you setting it?

**Comment by bghira** [comment]
and it already did the 100 epochs? we're just not detecting that?

**Comment by bghira** [comment]
ok, okay. that is different from my understanding. i'll have to reproduce this one later.

**Comment by bghira** [comment]
i wasn't able to reproduce this on the lastest release branch, it exports the pipeline and does final validations.

**Comment by bghira** [comment]
Fixed in #95
